### PR TITLE
Add Intel fast embeddings blog post

### PR DESCRIPTION
**Pull request by echarlaix**

### CODE REVIEW

File: intel-fast-embedding.md [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+---
+
+
+# CPU Optimized Embedding with 🤗 Optimum Intel and fastRAG
+
+Embedding models are useful for many applications such as retrieval, reranking, clustering, and classification. The research community has witnessed significant advancements in recent years in embedding models, leading to substantial enhancements in all applications building on semantic representation. Models such as [BGE](http://BAAI/bge-large-en-v1.5), [GTE](http://thenlper/gte-small), and [E5](http://intfloat/multilingual-e5-large) family of models are placed at the top of the [MTEB](https://github.com/embeddings-benchmark/mteb) benchmark and in some cases outperform proprietary embedding services. There are a variety of model sizes found in Hugging Face's Model hub, from lightweight (100-350M parameters) to 7B models (such as [Salesforce/SFR-Embedding-Mistral](http://Salesforce/SFR-Embedding-Mistral)). The lightweight models based on an encoder architecture are ideal candidates for optimization and utilization on CPU backends running semantic search-based applications, such as Retrieval Augmented Generation ([RAG](https://en.wikipedia.org/wiki/Prompt_engineering#Retrieval-augmented_generation)).
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:20]: nit: weird phrasing "Models such as ..." and then "E5 family of models" in the same sentence in my opinion.
```suggestion
Embedding models are useful for many applications such as retrieval, reranking, clustering, and classification. The research community has witnessed significant advancements in recent years in embedding models, leading to substantial enhancements in all applications building on semantic representation. Models such as [BGE](http://BAAI/bge-large-en-v1.5), [GTE](http://thenlper/gte-small), and the [E5](http://intfloat/multilingual-e5-large) family are placed at the top of the [MTEB](https://github.com/embeddings-benchmark/mteb) benchmark and in some cases outperform proprietary embedding services. There are a variety of model sizes found in Hugging Face Hub, from lightweight (100-350M parameters) to bigger (7B parameters) models, such as [Salesforce/SFR-Embedding-Mistral](http://Salesforce/SFR-Embedding-Mistral)) for instance. The lightweight models based on an encoder architecture are ideal candidates for optimization and utilization on CPU backends running semantic search-based applications, such as Retrieval Augmented Generation ([RAG](https://en.wikipedia.org/wiki/Prompt_engineering#Retrieval-augmented_generation)).
```

File: intel-fast-embedding.md [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+
+## Optimizing Embedding Models with Optimum Intel and IPEX
+
+[Optimum Intel](https://github.com/huggingface/optimum-intel) is an open-source library that accelerates end-to-end pipelines built with Hugging Face libraries on Intel Hardware. Optimum Intel includes several techniques to accelerate models such as low-bit quantization, model weight pruning, distillation, and an accelerated runtime.
+
+The runtime and optimizations included in [Optimum Intel](https://github.com/huggingface/optimum-intel) take advantage of Intel® Advanced Vector Extensions 512 (Intel® AVX-512) Vector Neural Network Instructions (VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX) on Intel CPUs to accelerate models. Specifically, it has built-in [BFloat16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) (BF16) and Int8 GEMM accelerators in every core to accelerate deep learning training and inference workloads. AMX accelerated inference is introduced through PyTorch 2.0 and [Intel Extension for PyTorch](https://github.com/intel/intel-extension-for-pytorch) (IPEX) in addition to other optimizations for various common operators.
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:52]:
```suggestion
The runtime and optimizations included in [Optimum Intel](https://github.com/huggingface/optimum-intel) take advantage of Intel® Advanced Vector Extensions 512 (Intel® AVX-512), Vector Neural Network Instructions (VNNI) and Intel® Advanced Matrix Extensions (Intel® AMX) on Intel CPUs to accelerate models. Specifically, it has built-in [BFloat16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) (BF16) and Int8 GEMM accelerators in every core to accelerate deep learning training and inference workloads. AMX accelerated inference is introduced in PyTorch 2.0 and [Intel Extension for PyTorch](https://github.com/intel/intel-extension-for-pytorch) (IPEX) in addition to other optimizations for various common operators.
```

File: intel-fast-embedding.md [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+Optimizing pre-trained models can be done easily with Optimum Intel; many simple examples can be found [here](https://huggingface.co/docs/optimum/main/en/intel/optimization_inc).
+
+
+## Example: Optimizing BGE Embedding Models
+
+In this blog, we focus on recently released embedding models by researchers at the [Beijing Academy of Artificial Intelligence](https://arxiv.org/pdf/2309.07597.pdf), as their models show competitive results on the widely adopted [MTEB](https://github.com/embeddings-benchmark/mteb) leaderboard.
+
+### BGE Technical Details
+
+Bi-encoder models are Transformer-based encoders trained to minimize a similarity metric, such as cosine-similarity, between two semantically similar texts as vectors. For example, popular Embedding models use a BERT model as a base pre-trained model and fine-tune it for embedding documents. The vector representing the encoded text is created from the model outputs; for example, it can be the [CLS] token vector or a mean of all the token vectors.
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:63]:
```suggestion
Bi-encoder models are Transformer-based encoders trained to minimize a similarity metric, such as cosine-similarity, between two semantically similar texts as vectors. For example, popular embedding models use a BERT model as a base pre-trained model and fine-tune it for embedding documents. The vector representing the encoded text is created from the model outputs; for example, it can be the [CLS] token vector or a mean of all the token vectors.
```

File: intel-fast-embedding.md [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+We focus on 3 BGE models: [small](https://huggingface.co/BAAI/bge-small-en-v1.5), [base](https://huggingface.co/BAAI/bge-base-en-v1.5), and [large](https://huggingface.co/BAAI/bge-large-en-v1.5) consisting of 45M, 110M, and 355M parameters encoding to 384/768/1024 sized embedding vectors, respectively.
+
+We note that the optimization process we showcase below is generic and can be applied to other embedding models (including bi-encoders, cross-encoders, and such).
+
+
+### Step-by-step: Optimization by Quantization
+
+We present a step-by-step guide for enhancing the performance of embedding models, focusing on reducing latency (with a batch size of 1) and increasing throughput (measured in documents encoded per second). This recipe utilizes optimum-intel and [Intel Neural Compressor](https://github.com/intel/neural-compressor) to quantize the model, and uses [IPEX](https://github.com/intel/intel-extension-for-pytorch) for optimized runtime on Intel-based hardware.
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:74]:
```suggestion
We present a step-by-step guide for enhancing the performance of embedding models, focusing on reducing latency (with a batch size of 1) and increasing throughput (measured in documents encoded per second). This recipe utilizes `optimum-intel` and [Intel Neural Compressor](https://github.com/intel/neural-compressor) to quantize the model, and uses [IPEX](https://github.com/intel/intel-extension-for-pytorch) for optimized runtime on Intel-based hardware.
```

File: intel-fast-embedding.md [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+```
+
+##### Step 2: Post-training Static Quantization
+
+Post-training static quantization requires a calibration set to determine the dynamic range of weights and activations. The calibration is done by running a representative set of data samples through the model, collecting statistics, and then quantizing the model based on the gathered info to minimize the accuracy loss.
+
+The following snippet shows a code snippet for quantization:
+
+```python
+def quantize(model_name: str, output_path: str, calibration_set: Dataset):
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:92]: nit: I would specify which kind of `Dataset` is expected here since it can be a `torch.utils.data.Dataset` or a `datasets.Dataset`.
```suggestion
def quantize(model_name: str, output_path: str, calibration_set: datasets.Dataset):
```

File: intel-fast-embedding.md [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+    # get the [CLS] token
+    embeddings = outputs[0][:, 0]
+```
+
+We provide additional and important details on how to configure the CPU-backend setup in the evaluation section below (correct machine setup).
+
+
+### Model Evaluation with MTEB
+
+Quantizing the models' weights to a lower precision introduces accuracy loss, as we lose precision moving from float32 weights to int8. Therefore, we aim to validate the accuracy of the optimized models by comparing them to the original models with two [MTEB](https://github.com/embeddings-benchmark/mteb)'s tasks:
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:145]:
```suggestion
Quantizing the models' weights to a lower precision introduces accuracy loss, as we lose precision moving from `float32` weights to `int8`. Therefore, we aim to validate the accuracy of the optimized models by comparing them to the original models with two [MTEB](https://github.com/embeddings-benchmark/mteb) tasks:
```

File: intel-fast-embedding.md [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+
+- **Retrieval**  - where a corpus is encoded and ranked lists are created by searching the index given a query.
+- **Reranking**  - reranking the retrieval's results for better relevance given a query.
+
+The table below shows the average accuracy (on multiple datasets)  of each task type (MAP for Reranking, NDCG@10 for Retrieval), where int8 is our quantized model and fp32 is the original model (results taken from the official MTEB leaderboard). The quantized models show less than 1% error rate compared to the original model in the Reranking task and less than 1.55% in the Retrieval task.
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:150]: nit: sometimes you use `fp32` and sometimes `float32`, I would use the same name.

File: intel-fast-embedding.md [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+| 0.5242 | 0.5325 | -1.55% |
+| 0.5346 | 0.5429 | -1.53% |
+
+</td></tr> </table>
+
+
+### Speed and Latency
+
+We compare the performance of our models with two other common methods of usage of models:
+1. Using PyTorch and Huggingface's Transformers library with bf16.
```
Comment (michaelbenayoun) [inline intel-fast-embedding.md:176]: Same comment for `bf16`

**Review by michaelbenayoun — APPROVED** [review]
LGTM
Left a few nits that you can feel free to ignore!

File: intel-fast-embedding.md (hunk 5 above)
Comment (echarlaix) [reply intel-fast-embedding.md:92]: good point !
